cfpb/cfgov-refresh
# eRegs progressive web app

This PR sets the foundation for service workers and general [PWA](https://developers.google.com/web/progressive-web-apps/checklist)-ness of the new eRegs. [sw-precache](https://github.com/GoogleChromeLabs/sw-precache) is used to generate a service worker that caches eRegs pages so that subsequent page loads are loaded from memory instead of the network. It should make the browsing of regulations super duper fast.

Service workers need to reside at the website's root (i.e. `https://consumerfinance.gov/service-worker.js`) so I named it `eregs-service-worker.js` to avoid any confusion. We may want to create new service workers for other parts of cf.gov in the future.

It's not picking up 100% of eReg's assets so the sw-precache config will need some further tweaking. You can verify the manifest and service worker are working by viewing the Application section of chrome dev tools:

![screen shot 2018-04-24 at 1 59 23 am](https://user-images.githubusercontent.com/1060248/39168942-cde3533e-4763-11e8-9a4c-e5892151e854.png)

![screen shot 2018-04-24 at 1 59 38 am](https://user-images.githubusercontent.com/1060248/39168946-d1e7a0ca-4763-11e8-8a2f-66b6242cc02e.png)

You can also verify that the `prerender` resource hint is successfully downloading the next reg section page in the background by visiting `chrome://net-internals/#prerender`.

![screen shot 2018-04-24 at 2 01 16 am](https://user-images.githubusercontent.com/1060248/39168933-c670269a-4763-11e8-8505-7f94c0ae6f14.png)

## Additions

- Custom webpack config file for eRegs
- Service worker generating script in the webpack file
- Unofficial-definitely-not-approved-but-still-cool eRegs icons designed by the amazing @niqjohnson 
- A [web app manifest](https://developers.google.com/web/fundamentals/web-app-manifest/) that tells browsers about the website

## Removals

- eRegs placeholder JS

## Testing

1. `./frontend.sh`
1. Load a new Regulations 3k reg section page.
1. Open the Application tab in chrome dev tools and see if the manifest and service worker are present.

## Todos

- Ensure all assets are being cached.
- Need to create a 200 offline page.
- The site's mobile experience needs to be improved in general.

## Screenshots

Android users will see a popup recommending they "install" eRegs by saving it to their home screen (please forgive the temporary awkward mobile menu).

![eregs-pwa](https://user-images.githubusercontent.com/1060248/39169414-ba7bb67c-4765-11e8-9a8e-cd28bd0154df.gif)
